### PR TITLE
Adding timer using jQuery, changing the handling ways in lesson controller

### PR DESCRIPTION
**Pull request by nhs3108**

### CODE REVIEW

File: app/models/lesson.rb [hunk 1]
```diff
@@ -7,7 +7,11 @@ class Lesson < ActiveRecord::Base
   accepts_nested_attributes_for :results
   after_save :save_activity
   def num_of_correct_ans
-    self.results.select{|result| result.answer.is_correct}.sum
+    self.results.select{|result| !result.answer.nil? && result.answer.is_correct}.length
+  end
+
+  def time_remaining
+    Settings.lesson.duration_in_minutes*60 - (Time.zone.now - self.updated_at).to_i
```
Comment (ThuBM) [inline app/models/lesson.rb]: Settings.lesson.duration_in_minutes.minutes

File: app/views/lessons/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      <% end %>
+      <div class="col-md-4 col-md-offset-4">
+        <br>
+        <%= f.submit t("done"), class: "btn btn-primary" %>
+      </div>
+    <% end %>
+  </div>
+</div>
+
+<script type="text/javascript">
```
Comment (ThuBM) [inline app/views/lessons/show.html.erb]: Viet trong file js

File: app/views/lessons/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      <div class="col-md-4 col-md-offset-4">
+        <br>
+        <%= f.submit t("done"), class: "btn btn-primary" %>
+      </div>
+    <% end %>
+  </div>
+</div>
+
+<script type="text/javascript">  
+  $(document).ready(time_remaining(<%= @time_remaining %>));
```
Comment (ThuBM) [inline app/views/lessons/show.html.erb]: Khong truyen tham so cua ruby trong javascript

File: app/views/lessons/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      $("#time-remaining").text("Time over");
+      if(total_seconds==0){
+        $("form").submit();
+      }
+      $("[type='submit']").remove()
+      return;
+    }else{
+      var minutes = Math.floor(total_seconds / 60)
+      var seconds = total_seconds % 60
+      display = minutes + " <%= Settings.unit.minute %> : "
```
Comment (ThuBM) [inline app/views/lessons/show.html.erb]: Tim hieu I18n.js

File: app/models/lesson.rb (hunk 1 above)
Comment (nhs3108) [inline app/models/lesson.rb]: I think it is not necessary

File: app/assets/javascripts/lesson_timer.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+function time_remaining(total_seconds, time_over_msg){
+  if(total_seconds<=0){
```
Comment (ThuBM) [inline app/assets/javascripts/lesson_timer.js]: indent

File: app/assets/javascripts/lesson_timer.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    $("#time-remaining").text(time_over_msg);
+    if(total_seconds==0){
```
Comment (ThuBM) [inline app/assets/javascripts/lesson_timer.js]: indent

File: app/assets/javascripts/lesson_timer.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+      $("form").submit();
+    }
+    $("[type='submit']").remove()
+    return;
+  }else{
+    var minutes = Math.floor(total_seconds / 60)
+    var seconds = total_seconds % 60
+    $("#time-remaining").text(minutes + " : " + seconds);
+    setTimeout(function(){time_remaining(total_seconds-1)}, 1000);
```
Comment (ThuBM) [inline app/assets/javascripts/lesson_timer.js]: indent

File: app/controllers/lessons_controller.rb [hunk 8]
```diff
@@ -1,40 +1,53 @@
...
-    if @lesson.save
-      redirect_to category_lesson_path @category, @lesson
+  def show
+    @time_remaining = @lesson.time_remaining
+    @is_time_over = @time_remaining < 0 ? true : false
+    @mark = @lesson.num_of_correct_ans if @is_time_over
+  end
+
+  def update
+    if @lesson.update lesson_params
```
Comment (ThuBM) [inline app/controllers/lessons_controller.rb]: find @lesson by params[:id]???

File: app/controllers/lessons_controller.rb [hunk 9]
```diff
@@ -1,40 +1,53 @@
...
       flash[:danger] = t "not_enough_word"
-      redirect_to :back
+      redirect_to categories_path
     else
-      @lesson = @category.lessons.build user: current_user
-      @words.each do |word|
-        result = @lesson.results.build
-        result.word = word
+      @lesson = Lesson.create category_id: @category.id, user_id: current_user.id
+      @words.limit(Settings.lesson.question_limit_words).each do |word|
```
Comment (chuanhtuan) [inline app/controllers/lessons_controller.rb]: load toàn bộ word lên rồi mới limit
đoạn này chạy khá chậm

**Comment by ThuBM** [comment]
OK

File: app/controllers/lessons_controller.rb [hunk 10]
```diff
@@ -1,40 +1,53 @@
...
       flash[:danger] = t "not_enough_word"
-      redirect_to :back
+      redirect_to categories_path
     else
-      @lesson = @category.lessons.build user: current_user
+      @lesson = Lesson.create category_id: @category.id, user_id: current_user.id
       @words.each do |word|
-        result = @lesson.results.build
-        result.word = word
+        Result.create(lesson_id: @lesson.id, word_id: word.id)
```
Comment (dieunb) [inline app/controllers/lessons_controller.rb]: move add words to lesson to call_back

File: app/views/lessons/show.html.erb [hunk 11]
```diff
@@ -16,3 +20,7 @@
     <% end %>
   </div>
 </div>
+
+<script type="text/javascript">
```
Comment (dieunb) [inline app/views/lessons/show.html.erb]: move js code to js file

File: app/controllers/lessons_controller.rb [hunk 12]
```diff
@@ -1,40 +1,51 @@
...
-      redirect_to :back
+      redirect_to categories_path
     else
-      @lesson = @category.lessons.build user: current_user
-      @words.each do |word|
-        result = @lesson.results.build
-        result.word = word
-      end
+      @lesson = Lesson.create category_id: @category.id, user_id: current_user.id
+      @lesson.init_result @words
```
Comment (dieunb) [inline app/controllers/lessons_controller.rb]: call_back

File: app/controllers/lessons_controller.rb [hunk 13]
```diff
@@ -1,40 +1,50 @@
...
 
-  def create
-    @category = Category.find params[:category_id]
-    @lesson = @category.lessons.build lesson_params
-    @lesson.user_id = current_user.id
-    if @lesson.save
-      redirect_to category_lesson_path @category, @lesson
+  def show
+    @time_remaining = @lesson.time_remaining
+    @is_time_over = @time_remaining < 0 ? true : false
```
Comment (dieunb) [inline app/controllers/lessons_controller.rb]: just `@is_time_over = @time_remaining < 0`

File: app/models/lesson.rb [hunk 14]
```diff
@@ -6,8 +6,22 @@ class Lesson < ActiveRecord::Base
...
+  end
+
+  def time_remaining
+    Settings.lesson.duration_in_minutes*60 - (Time.zone.now - self.created_at).to_i
+  end
+
+  private
+  def init_result
+    words = self.category.words.not_learned(self.user).random_questions
+    words.each do |word|
```
Comment (dieunb) [inline app/models/lesson.rb]: one line of code and why not `self.result.create word_id: word.id`

**Pull request merged**